### DISCUSSION TITLE
Add option to control bandwidth usage during cloud backup/restore operations

### DISCUSSION THREAD

**Discussion by OliverHfm** [cloudnative-pg/cloudnative-pg#1327] [Ideas]
We noticed that CNPG operator is struggling in backup/restore situations if our implemented cloud storage bandwidth limits are hit/exceeded by CNPG. 

E.g.
- CNPG backup to S3 storage fails if the bandwidth limit is exceeded
- CNPG cluster initialisation from backup fails with following message in the logs

`ERROR: Barman cloud restore exception: An error occurred (SlowDown) when calling the GetObject operation (reached max retries: 4): Please reduce your request rate.`

We think adding the option to limit bandwidth usage during backup/restore will be very beneficial to control bandwidth usage.
E.g. if the database only reads/writes from/to S3 with up to a given bandwidth an admin could ensure the operation is not exceeding any implemented QoS limits for the network storage. This in return would prevent CNPG hitting any related network traffic error.

From an internet search some mechanism for traffic control like 'trickle' or 'pv' might be options built a feature to control the backup/restore related traffic:  
- https://github.com/aws/aws-cli/issues/1090)
- https://linux.die.net/man/1/pv

**Comment by gbartolini** [comment]
Barman cloud does not allow us to specify a bandwidth limitation because the S3 protocol does not enable control of that option, AFAIK.

I think a more interesting approach would be to explore the bandwidth control capability made available by the CNI, as explained here:

https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/

This could be done for the recovery job and the backup jobs, through the `kubernetes.io/ingress-bandwidth` and `kubernetes.io/egress-bandwidth` annotations, presumably. Would that help?

Any thoughts from the other maintainers?

**Comment by sxd** [comment]
Indeed the best idea will be use the CNI capabilities here to restrict the bandwidth if that's what you need to, but also, increase the bandwidth limit on the S3 side.

Another possible approach here, will be to have MinIO locally deployed and then make MinIO able to replicate to the S3 object storage if it's required to have that in the S3 object storage https://min.io/docs/minio/linux/administration/object-management/transition-objects-to-s3.html

I think that any bandwidth limit shouldn't be applied on the user but at network administration level, that's why networking implements that capabilities, that's part of the CNI plugins or any network admin that can restrict the limits to certain origin/destination.

**Comment by gbartolini** [comment]
As a workaround, you could try and follow the instructions provided here: https://cloudnative-pg.io/documentation/current/labels_annotations/

You could add both `kubernetes.io/ingress-bandwidth` and `kubernetes.io/egress-bandwidth` annotations in the `INHERITED_ANNOTATIONS` environment variable of the operator's configuration, then add the annotations and desired values for the entire cluster.

**Comment by mikewallace1979** [comment]
From the Barman perspective it is possible to implement some throttling upstream of the S3 client, such as limiting the number of concurrent part uploads, however this would be pretty crude compared to an infrastructure level solution.

**Comment by OliverHfm** [comment]
Hi @gbartolini 

Barman team added '--max-bandwidth' option in v3.10.0 to allow manual bandwidth control.
Can this be exposed to the CNPG operator to make use of this with CNPG?

Regards
Oliver
